src/Stack/Upload.hs

```diff
@@ -67,12 +75,31 @@ instance FromJSON (FilePath -> HackageCreds) where
 withEnvVariable :: Text -> IO Text -> IO Text
 withEnvVariable varName fromPrompt = lookupEnv (T.unpack varName) >>= maybe fromPrompt (pure . T.pack)
 
+maybeGetHackageKey :: IO (Maybe String)
+maybeGetHackageKey = lookupEnv (T.unpack "HACKAGE_KEY")
+
+getCredsWithApiKey :: String -> FilePath -> IO HackageCreds
+getCredsWithApiKey key fp = return HackageCreds {
+    hcUsername = ""
+  , hcPassword = fromString key
+  , hcCredsFile = fp
+}
+
+loadCreds :: Config -> IO HackageCreds
+loadCreds config = do
+  maybeHackageKey <- maybeGetHackageKey
+  case maybeHackageKey of
+    Just key -> do
+      putStrLn "HACKAGE_KEY found in env, using that for credentials."
```

Comment:
should probably be `logInfo`?

Reply:
Right. I tried that but hit an error I couldn't solve. Perhaps you could help me get this straight. The error was:

```haskell
No instance for (MonadReader env0 IO)
        arising from a use of ‘logInfo’
-- (...)
93 |       logInfo "HACKAGE_KEY found in env, using that for credentials."
```

1. Importing `RIO`, as described [here](https://www.fpcomplete.com/haskell/library/rio/) didn't help,
2. Stack.Prelude is already imported in `Upload`, as in some other modules using `logInfo`,
3. `test/integration/lib/StackTest.hs` defines its own internal `logInfo`, but I suppose that's not the approach you have in mind here.

Reply:
```diff
diff --git a/src/Stack/Upload.hs b/src/Stack/Upload.hs
index 66ba4102..1b8918df 100644
--- a/src/Stack/Upload.hs
+++ b/src/Stack/Upload.hs
@@ -49,9 +49,11 @@ import           System.Directory                      (createDirectoryIfMissing
                                                         removeFile, renameFile)
 import           System.Environment                    (lookupEnv)
 import           System.FilePath                       ((</>), takeFileName, takeDirectory)
-import           System.IO                             (putStrLn, putStr, print) -- TODO remove putStrLn, use logInfo
 import           System.PosixCompat.Files              (setFileMode)
 
+
+newtype HackageKey = HackageKey Text
+
 -- | Username and password to log into Hackage.
 --
 -- Since 0.1.0.0
@@ -62,6 +64,9 @@ data HackageCreds = HackageCreds
     }
     deriving Show
 
+data HackageAuth = HAKey HackageKey
+                 | HACreds HackageCreds
+
 instance ToJSON HackageCreds where
     toJSON (HackageCreds u p _) = object
         [ "username" .= u
@@ -75,34 +80,27 @@ instance FromJSON (FilePath -> HackageCreds) where
 withEnvVariable :: Text -> IO Text -> IO Text
 withEnvVariable varName fromPrompt = lookupEnv (T.unpack varName) >>= maybe fromPrompt (pure . T.pack)
 
-maybeGetHackageKey :: IO (Maybe String)
-maybeGetHackageKey = lookupEnv (T.unpack "HACKAGE_KEY")
+maybeGetHackageKey :: RIO m (Maybe HackageKey)
+maybeGetHackageKey = fmap (HackageKey . T.pack) <$> (liftIO $ lookupEnv (T.unpack "HACKAGE_KEY"))
 
-getCredsWithApiKey :: String -> FilePath -> HackageCreds
-getCredsWithApiKey key fp = HackageCreds {
-    hcUsername = ""
-  , hcPassword = T.pack key
-  , hcCredsFile = fp
-}
 
-loadCreds :: Config -> IO HackageCreds
+loadCreds :: HasLogFunc m => Config -> RIO m HackageAuth
 loadCreds config = do
   maybeHackageKey <- maybeGetHackageKey
   case maybeHackageKey of
     Just key -> do
-      putStrLn "HACKAGE_KEY found in env, using that for credentials."
-      fp <- credsFile config
-      return $ getCredsWithApiKey key fp
-    Nothing -> loadUserAndPassword config
+      logInfo "HACKAGE_KEY found in env, using that for credentials."
+      return $ HAKey key
+    Nothing -> HACreds <$> loadUserAndPassword config
 
 -- | Load Hackage credentials, either from a save file or the command
 -- line.
 --
 -- Since 0.1.0.0
-loadUserAndPassword :: Config -> IO HackageCreds
+loadUserAndPassword :: HasLogFunc m => Config -> RIO m HackageCreds
 loadUserAndPassword config = do
-  fp <- credsFile config
-  elbs <- tryIO $ L.readFile fp
+  fp <- liftIO $ credsFile config
+  elbs <- liftIO $ tryIO $ L.readFile fp
   case either (const Nothing) Just elbs >>= \lbs -> (lbs, ) <$> decode' lbs of
     Nothing -> fromPrompt fp
     Just (lbs, mkCreds) -> do
@@ -111,14 +109,15 @@ loadUserAndPassword config = do
       writeFilePrivate fp $ lazyByteString lbs
 
       unless (configSaveHackageCreds config) $ do
-        putStrLn "WARNING: You've set save-hackage-creds to false"
-        putStrLn "However, credentials were found at:"
-        putStrLn $ "  " ++ fp
+        logInfo "WARNING: You've set save-hackage-creds to false"
+        logInfo "However, credentials were found at:"
+        logInfo $ "  " <> fromString fp
       return $ mkCreds fp
   where
+    fromPrompt :: HasLogFunc m => FilePath -> RIO m HackageCreds
     fromPrompt fp = do
-      username <- withEnvVariable "HACKAGE_USERNAME" (prompt "Hackage username: ")
-      password <- withEnvVariable "HACKAGE_PASSWORD" (promptPassword "Hackage password: ")
+      username <- liftIO $ withEnvVariable "HACKAGE_USERNAME" (prompt "Hackage username: ")
+      password <- liftIO $ withEnvVariable "HACKAGE_PASSWORD" (promptPassword "Hackage password: ")
       let hc = HackageCreds
             { hcUsername = username
             , hcPassword = password
@@ -128,10 +127,10 @@ loadUserAndPassword config = do
       when (configSaveHackageCreds config) $ do
         shouldSave <- promptBool $ T.pack $
           "Save hackage credentials to file at " ++ fp ++ " [y/n]? "
-        putStrLn "NOTE: Avoid this prompt in the future by using: save-hackage-creds: false"
+        logInfo "NOTE: Avoid this prompt in the future by using: save-hackage-creds: false"
         when shouldSave $ do
           writeFilePrivate fp $ fromEncoding $ toEncoding hc
-          putStrLn "Saved!"
+          logInfo "Saved!"
           hFlush stdout
 
       return hc
@@ -164,30 +163,30 @@ credsFile config = do
     createDirectoryIfMissing True dir
     return $ dir </> "credentials.json"
 
-addAPIKey :: String -> Request -> Request
-addAPIKey key req =
-  setRequestHeader "Authorization" [fromString $ "X-ApiKey" ++ " " ++ key] req
+addAPIKey :: HackageKey -> Request -> Request
+addAPIKey (HackageKey key) req =
+  setRequestHeader "Authorization" [fromString $ "X-ApiKey" ++ " " ++ T.unpack key] req
 
-applyKeyOrCreds :: HackageCreds -> Request -> IO Request
-applyKeyOrCreds creds req0 = do
-    case hcUsername creds of
-        "" -> return (addAPIKey (T.unpack $ hcPassword creds) req0)
-        _ -> applyCreds creds req0
+applyKeyOrCreds :: HasLogFunc m => HackageAuth -> Request -> RIO m Request
+applyKeyOrCreds haAuth req0 = do
+    case haAuth of
+        HAKey key -> return (addAPIKey key req0)
+        HACreds creds -> applyCreds creds req0
 
-applyCreds :: HackageCreds -> Request -> IO Request
+applyCreds :: HasLogFunc m => HackageCreds -> Request -> RIO m Request
 applyCreds creds req0 = do
-  manager <- getGlobalManager
-  ereq <- applyDigestAuth
+  manager <- liftIO $ getGlobalManager
+  ereq <- liftIO $ applyDigestAuth
     (encodeUtf8 $ hcUsername creds)
     (encodeUtf8 $ hcPassword creds)
     req0
     manager
   case ereq of
       Left e -> do
-          putStrLn "WARNING: No HTTP digest prompt found, this will probably fail"
+          logInfo "WARNING: No HTTP digest prompt found, this will probably fail"
           case fromException e of
-              Just e' -> putStrLn $ displayDigestAuthException e'
-              Nothing -> print e
+              Just e' -> logInfo $ fromString (displayDigestAuthException e')
+              Nothing -> logWarn $ fromString $ displayException e
           return req0
       Right req -> return req
 
@@ -195,13 +194,14 @@ applyCreds creds req0 = do
 -- sending a file like 'upload', this sends a lazy bytestring.
 --
 -- Since 0.1.2.1
-uploadBytes :: String -- ^ Hackage base URL
-            -> HackageCreds
+uploadBytes :: HasLogFunc m
+            => String -- ^ Hackage base URL
+            -> HackageAuth
             -> String -- ^ tar file name
             -> UploadVariant
             -> L.ByteString -- ^ tar file contents
-            -> IO ()
-uploadBytes baseUrl creds tarName uploadVariant bytes = do
+            -> RIO m ()
+uploadBytes baseUrl auth tarName uploadVariant bytes = do
     let req1 = setRequestHeader "Accept" ["text/plain"]
                (fromString $ baseUrl
                           <> "packages/"
@@ -210,31 +210,36 @@ uploadBytes baseUrl creds tarName uploadVariant bytes = do
                                Candidate -> "candidates/"
                )
         formData = [partFileRequestBody "package" tarName (RequestBodyLBS bytes)]
-    req2 <- formDataBody formData req1
-    req3 <- applyKeyOrCreds creds req2
-    putStr $ "Uploading " ++ tarName ++ "... "
+    req2 <- liftIO $ formDataBody formData req1
+    req3 <- applyKeyOrCreds auth req2
+    logInfo $ "Uploading " <> fromString tarName <> "... "
     hFlush stdout
-    withResponse req3 $ \res ->
+    withRunInIO $ \runInIO -> withResponse req3 (runInIO . inner)
+ where
+    inner :: HasLogFunc m => Response (ConduitM () S.ByteString IO ()) -> RIO m ()
+    inner res =
         case getResponseStatusCode res of
-            200 -> putStrLn "done!"
+            200 -> logInfo "done!"
             401 -> do
-                putStrLn "authentication failure"
-                handleIO (const $ return ()) (removeFile (hcCredsFile creds))
+                logError "authentication failure"
+                case auth of
+                  HACreds creds -> handleIO (const $ return ()) (liftIO $ removeFile (hcCredsFile creds))
+                  _ -> pure ()
                 throwString "Authentication failure uploading to server"
             403 -> do
-                putStrLn "forbidden upload"
-                putStrLn "Usually means: you've already uploaded this package/version combination"
-                putStrLn "Ignoring error and continuing, full message from Hackage below:\n"
-                printBody res
+                logError "forbidden upload"
+                logError "Usually means: you've already uploaded this package/version combination"
+                logError "Ignoring error and continuing, full message from Hackage below:\n"
+                liftIO $ printBody res
             503 -> do
-                putStrLn "service unavailable"
-                putStrLn "This error some times gets sent even though the upload succeeded"
-                putStrLn "Check on Hackage to see if your pacakge is present"
-                printBody res
+                logError "service unavailable"
+                logError "This error some times gets sent even though the upload succeeded"
+                logError "Check on Hackage to see if your pacakge is present"
+                liftIO $ printBody res
             code -> do
-                putStrLn $ "unhandled status code: " ++ show code
-                printBody res
-                throwString $ "Upload failed on " ++ tarName
+                logError $ "unhandled status code: " <> fromString (show code)
+                liftIO $ printBody res
+                throwString $ "Upload failed on " <> fromString tarName
 
 printBody :: Response (ConduitM () S.ByteString IO ()) -> IO ()
 printBody res = runConduit $ getResponseBody res .| CB.sinkHandle stdout
@@ -242,20 +247,22 @@ printBody res = runConduit $ getResponseBody res .| CB.sinkHandle stdout
 -- | Upload a single tarball with the given @Uploader@.
 --
 -- Since 0.1.0.0
-upload :: String -- ^ Hackage base URL
-       -> HackageCreds
+upload :: HasLogFunc m
+       => String -- ^ Hackage base URL
+       -> HackageAuth
        -> FilePath
        -> UploadVariant
-       -> IO ()
-upload baseUrl creds fp uploadVariant =
-  uploadBytes baseUrl creds (takeFileName fp) uploadVariant =<< L.readFile fp
+       -> RIO m ()
+upload baseUrl auth fp uploadVariant =
+  uploadBytes baseUrl auth (takeFileName fp) uploadVariant =<< (liftIO $ L.readFile fp)
 
-uploadRevision :: String -- ^ Hackage base URL
-               -> HackageCreds
+uploadRevision :: HasLogFunc m
+               => String -- ^ Hackage base URL
+               -> HackageAuth
                -> PackageIdentifier
                -> L.ByteString
-               -> IO ()
-uploadRevision baseUrl creds ident@(PackageIdentifier name _) cabalFile = do
+               -> RIO m ()
+uploadRevision baseUrl auth ident@(PackageIdentifier name _) cabalFile = do
   req0 <- parseRequest $ concat
     [ baseUrl
     , "package/"
@@ -269,5 +276,5 @@ uploadRevision baseUrl creds ident@(PackageIdentifier name _) cabalFile = do
     , partBS "publish" "on"
     ]
     req0
-  req2 <- applyKeyOrCreds creds req1
+  req2 <- applyKeyOrCreds auth req1
   void $ httpNoBody req2
diff --git a/src/main/Main.hs b/src/main/Main.hs
index bb0f2287..9234f71c 100644
--- a/src/main/Main.hs
+++ b/src/main/Main.hs
@@ -655,21 +655,19 @@ uploadCmd uploadOpts = do
         config <- view configL
         let hackageUrl = T.unpack $ configHackageBaseUrl config
             uploadVariant = uoptsUploadVariant uploadOpts
-        getCreds <- liftIO $ memoizeRef $ Upload.loadCreds config
+        getCreds <- memoizeRef $ Upload.loadCreds config
         mapM_ (resolveFile' >=> checkSDistTarball sdistOpts) files
         forM_ files $ \file -> do
             tarFile <- resolveFile' file
-            liftIO $ do
-              creds <- runMemoized getCreds
-              Upload.upload hackageUrl creds (toFilePath tarFile) uploadVariant
+            creds <- runMemoized getCreds
+            Upload.upload hackageUrl creds (toFilePath tarFile) uploadVariant
         forM_ dirs $ \dir -> do
             pkgDir <- resolveDir' dir
             (tarName, tarBytes, mcabalRevision) <- getSDistTarball (sdoptsPvpBounds sdistOpts) pkgDir
             checkSDistTarball' sdistOpts tarName tarBytes
-            liftIO $ do
-              creds <- runMemoized getCreds
-              Upload.uploadBytes hackageUrl creds tarName uploadVariant tarBytes
-              forM_ mcabalRevision $ uncurry $ Upload.uploadRevision hackageUrl creds
+            creds <- runMemoized getCreds
+            Upload.uploadBytes hackageUrl creds tarName uploadVariant tarBytes
+            forM_ mcabalRevision $ uncurry $ Upload.uploadRevision hackageUrl creds
 
 sdistCmd :: SDistOpts -> RIO Runner ()
 sdistCmd sdistOpts =
```